Dockerfile

```diff
@@ -44,57 +42,42 @@ ARG EXTRA_PACKAGES_VERSION=master
 ENV EXTRA_PACKAGES_VERSION ${EXTRA_PACKAGES_VERSION}
 
 # Set up the SSH key for usage by git
-RUN mkdir /root/.ssh/
-RUN echo "${SSH_PRIVATE_KEY}" > /root/.ssh/id_rsa
-RUN chmod 600 /root/.ssh/id_rsa
-RUN touch /root/.ssh/known_hosts
-RUN ssh-keyscan github.com >> /root/.ssh/known_hosts
+RUN mkdir ~/.ssh/ && \
+        echo "${SSH_PRIVATE_KEY}" > ~/.ssh/id_rsa && \
+        chmod 600 ~/.ssh/id_rsa && \
+        touch ~/.ssh/known_hosts && \
+        ssh-keyscan github.com >> ~/.ssh/known_hosts
 
 # Acquire the software source
-COPY . /root/src/CARMAPlatform/
-RUN /root/src/CARMAPlatform/docker/checkout.sh
-
+RUN mkdir ~/src
+COPY --chown=carma . /home/carma/src/CARMAPlatform/
+RUN ~/src/CARMAPlatform/docker/checkout.sh
```

Comment:
Not sure I follow what these lines are doing. It appears the assumption is this is run by a user named "carma".  If any other user then line 53 will be populating a different directory than the one created in line 52.  Also, not clear what is intended to be copied in line 53.  Copying the content of . will recursively grab possibly all kinds of garbage.

Reply:
Correct, these commands are running inside the docker container as it's being built. The line `USER carma` in the base image defines that all future commands are to be run as the carma user, ensuring that this is the case. Since `RUN` invokes bash to process the command, `~` resolves correctly to the `/home/carma` folder inside the image, the `COPY` command is resolved by the Docker daemon itself with no bash processing, so I have to spell out the directory in full.

The purpose of copying the content of `.` as it exists is to allow for developers to build their current working tree into a docker image. This is relatively standard practice as far as I know, if you want to build an image for a specific version of your software you check that version out and it gets copied in during the build process or if you want to build your local changes those get copied in as well. Any garbage is filtered out later, the final image contains on the compiled artifacts to save on space. If you look at the stages, all contents from the previous stage is totally discarded unless copied over by a `COPY --from=<STAGE_NAME>` command.